lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/cache/EvictableCache.java

```diff
@@ -0,0 +1,194 @@
+ * but one that does not exhibits <a href="https://github.com/google/guava/issues/1881">Guava issue #1881</a>, i.e.
+ * a {@link #getIfPresent(Object)} after {@link #invalidate(Object)} is guaranteed to return {@code null} and
+ * {@link #get(Object, Callable)} after {@link #invalidate(Object)} is guaranteed to load a fresh value.
+ */
+public class EvictableCache<K, V>
+        extends AbstractCache<K, V>
+        implements Cache<K, V>
+{
+    /**
+     * @apiNote Piggy-back on {@link CacheBuilder} for cache TTL.
+     */
+    public static <K, V> EvictableCache<K, V> buildWith(CacheBuilder<? super K, Object> cacheBuilder)
+    {
+        return new EvictableCache<>(cacheBuilder);
+    }
+
+    // private final Map<K, Future<V>> map = new ConcurrentHashMap<>();
+    private final Cache<K, Future<V>> delegate;
+
+    private final StatsCounter statsCounter = new SimpleStatsCounter();
+
+    public EvictableCache(CacheBuilder<? super K, Object> cacheBuilder)
+    {
+        this.delegate = cacheBuilder.build();
+    }
+
+    @CheckForNull
+    @Override
+    public V getIfPresent(Object key)
+    {
+        Future<V> future = delegate.getIfPresent(key);
+        if (future != null && future.isDone()) {
+            statsCounter.recordHits(1);
+            return getDone(future);
+        }
+        return null;
+    }
+
+    @Override
+    public V get(K key, Callable<? extends V> loader)
+            throws ExecutionException
+    {
+        requireNonNull(key, "key is null");
+        requireNonNull(loader, "loader is null");
+
+        while (true) {
+            SettableFuture<V> newFuture = SettableFuture.create();
+            Future<V> future = delegate.asMap().computeIfAbsent(key, ignored -> newFuture);
+            if (future.isDone() && !future.isCancelled()) {
+                statsCounter.recordHits(1);
+                return getDone(future);
+            }
+
+            statsCounter.recordMisses(1);
+            if (future == newFuture) {
+                // We put the future in.
+
+                V computed;
+                long loadStartNanos = nanoTime();
+                try {
+                    computed = loader.call();
+                    requireNonNull(computed, "computed is null");
+                }
+                catch (Exception e) {
+                    statsCounter.recordLoadException(nanoTime() - loadStartNanos);
+                    delegate.asMap().remove(key, newFuture);
+                    // wake up waiters, let them retry
+                    newFuture.cancel(false);
+                    throw new ExecutionException(e);
+                }
+                statsCounter.recordLoadSuccess(nanoTime() - loadStartNanos);
+                newFuture.set(computed);
+                return computed;
+            }
+
+            // Someone else is loading the key, let's wait.
+            try {
+                // TODO should this record hit or miss?
```

Comment:
I would say miss. It is safer to be conservative when measuring cache efficency.

Reply:
You already recorded this as miss above.

Reply:
I know. Thanks. Removing the TODO.